## # 1.字符串截取的易错点
解析：由于go语言中的字符串实际上是类型为byte的只读切片。或者说一个字符串就是一堆字节。这意味着，当我们将字符存储在字符串中时，实际存储的是这个字符的字节。一个字符串包含了任意个byte，它并不限定Unicode，UTF-8或者任何其他预定义的编码。
```go
func main() {
	x := "text"
	y := "四维合创"
	fmt.Println("打印x",x[0])
	fmt.Println("打印x",x[0:1])
	fmt.Println("打印y",y[0])
	fmt.Println("打印y",y[0:1])
	fmt.Println("打印y",y[0:3])
}
```
## # 2.可变参数
解析：从内部实现机理上来说，类型…type本质上是一个数组切片，也就是[]type
```go
func main() {
	fmt.Println("aaaa",add(1, 2))
	fmt.Println("bbbb",add([]int{1, 3, 7}...))
}


func add(args ...int) int {
	sum := 0
	for _,arg := range args {
		sum += arg
	}
	return sum
}
```

## # 3.Panic
1.这里fc为nil，因而产生runtime panic，然后被defer的recover捕获
```go
var fc func() string

func protect(g func() string) {
	defer func() {
		fmt.Println("done") 
		if x := recover(); x != nil {
			fmt.Printf("run time panic: %v", x)
		}
	}()
	fmt.Println("start")
	fmt.Println(g())
}
func main() {
	protect(fc)
}
```
2.这里defer2捕获了panic，defer1就捕获不到panic了
```go
func deferRecover() {
	defer func() {
		fmt.Println("defer1: recover")
		if err := recover(); err != nil {
			fmt.Println(err)
		} else {
			fmt.Println("defer1 recover nil")
		}
	}()

	defer func() {
		fmt.Println("defer2: recover ")
		if err := recover(); err != nil {
			fmt.Println(err)
		}
	}()

	panic("manual panic")
}
```
3.这里defer2产生了panic，"覆盖"了原始的panic，defer1捕获的是最后一个panic即defer2产生的panic
```go
func paincInDefer() {
    defer func() {
        fmt.Println("defer1 recover")
        if err := recover(); err != nil {
            // main panic is override by defer2 panic
            fmt.Println("aaaaa",err)
        } else {
            fmt.Println("defer1 recover nil")
        }
    }()

    defer func() {
        fmt.Println("defer2 begin to panic")
        panic("defer2 panic")
    }()

    panic("main panic")
}
```
4.这里defer2产生了panic，defer1没有捕获，产生的panic依次显示了main panic及defer2 panic
```go
func paincInDeferNotRecover() {
	defer func() {
		fmt.Println("defer1")
	}()

	defer func() {
		fmt.Println("defer2 begin to panic")
		panic("defer2 panic")
	}()

	panic("main panic")
}
```
## # 4.http请求问题
解析：虽然执行了 6 次循环，而且每次都没有执行 Body.Close() ,就是因为执行了ioutil.ReadAll()把内容都读出来了，连接得以复用，因此只泄漏了一个读goroutine和一个写goroutine，最后加上main goroutine，所以答案就是3个goroutine。从另外一个角度说，正常情况下我们的代码都会执行 ioutil.ReadAll()，但如果此时忘了 resp.Body.Close()，确实会导致泄漏。但如果你调用的域名一直是同一个的话，那么只会泄漏一个 读goroutine 和一个写goroutine，这就是为什么代码明明不规范但却看不到明显内存泄漏的原因。
```go
func main() {
	num := 6
	for index := 0; index < num; index++ {
		resp, _ := http.Get("https://www.baidu.com")
		_, _ = ioutil.ReadAll(resp.Body)
	}
	fmt.Printf("此时goroutine个数= %d\n", runtime.NumGoroutine())
}
```
## # 5.go-zero
链接：https://gitee.com/kevwan/go-zero